content/Phoenix for Rails developers - Bruno Vezoli.md

```diff
@@ -0,0 +1,347 @@
+```
+
+```bash
+mix phx.gen.html Articles Article articles title:string text:text
+```
+
+Once more they look alike with subtle differences, in this case Phoenix goes for a more verbose fashion by asking us to specify the context module where it'll be created, the name of the schema module (equivalent to the model) and it's plural to be used as the table name.
+
+### Migrations
+
+Speaking of table names, let's take a look at the way migrations are done. First, to refresh the memory, ActiveRecord's migrations:
+
+```ruby
+def change
+  create_table :articles do |t|
+    t.string :title
+    t.text :text
+
+    t.timestamps
+  end
+end
+```
+And now Ecto's (Phoenix's ORM):
+
+```elixir
+def change do
+  create table(:articles) do
+    add :title, :string
+    add :text, :text
+
+    timestamps()
+  end
+end
+```
+
+Notice the difference? At first glance it looks like the same thing, but if you take a closer look you'll notice that the first snippet takes an Object Oriented approach while the second one is functional. This makes it very easy to apply our Rails knowledge into this new and exciting framework, we just need to tweak a bit the syntax but the basis is still there. For example, let's add a unique constraint to our articles' title.
+
+```elixir
+def change do
+  create table(:articles) do
+    add :title, :string
+    add :text, :text
+
+    timestamps()
+  end
+
+  create unique_index(:articles, [:title])
+end
+```
+
+So even though it might be different to the common approach of adding the `index: { unique: true }`, it's still very similar to the way an index is created on an existing table. Mainly because that's what's happening here.
+
+### Models and schemas
+
+Here we find one of the first great differences. While models are the heart of a Rails app, Phoenix just has schemas. These are basically structs which contain the database data, like the table fields and their types. It is also responsible of validating this data and generating error messages. Aside from that it doesn't know a thing about the database, it does not save update or delete any records. But enough chatter, let's see how they look:
+
+```elixir
+defmodule Blog.Articles.Article do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "articles" do
+    field :text, :string
+    field :title, :string
+
+    timestamps()
+  end
+
+  def changeset(article, attrs) do
+    article
+    |> cast(attrs, [:title, :text])
+    |> validate_required([:title, :text])
+    |> unique_constraint(:title)
+  end
+end
+```
+
+Don't be scared of all those `|>`, it's just the pipe operator (just like Unix's). It just passes the result on the left-hand side as the first argument of the function in the right-hand side. So for example cast is being called with `article` as it's first argument, making the two first lines equivalent to `cast(article, attrs, [:title, :text])`. 
+
+Phoenix's solution is once again it's much more verbose than your standard ActiveRecord:
```

Comment:
`Phoenix's solution is once again much more verbose than your standard ActiveRecord`